include/dslash_helper.cuh

```diff
@@ -601,70 +611,94 @@ namespace quda
+    static constexpr bool dagger = dagger_;
+    static constexpr bool xpay = xpay_;
+    static constexpr KernelType kernel_type = kernel_type_;
+    Arg arg;
+
+    dslash_functor_arg(const Arg &arg, unsigned int threads_x) :
+      kernel_param(dim3(threads_x, arg.dc.Ls, arg.nParity)),
+      arg(arg) { }
+  };
 
-    if ((kernel_type == INTERIOR_KERNEL || kernel_type == UBER_KERNEL) && blockIdx.x < arg.pack_blocks) {
-      // first few blocks do packing kernel
-      P<dagger, dslash.pc_type(), Arg> packer;
-      packer(arg, s, 1 - parity, dslash.twist_pack()); // flip parity since pack is on input
+  /**
+    @brief This is the functor for the dslash stencils.
 
-      // we use that when running the exterior -- this is either
-      // * an explicit call to the exterior when not merged with the interior or
-      // * the interior with exterior_blocks > 0
+    When running an interior kernel, the first few "pack_blocks" CTAs
+    are reserved for data packing, which may include communication to
+    neighboring processes.
+   */
+  template <typename Arg> struct dslash_functor {
+    const typename Arg::Arg &arg;
+    static constexpr int nParity = Arg::nParity;
+    static constexpr bool dagger = Arg::dagger;
+    static constexpr KernelType kernel_type = Arg::kernel_type;
+    static constexpr const char *filename() { return Arg::D::filename(); }
+    constexpr dslash_functor(const Arg &arg) : arg(arg.arg) { }
+
+    __forceinline__ __device__ void operator()(int, int s, int parity)
+    {
+      typename Arg::D dslash(arg);
+      // for full fields set parity from z thread index else use arg setting
+      if (nParity == 1) parity = arg.parity;
+
+      if ((kernel_type == INTERIOR_KERNEL || kernel_type == UBER_KERNEL) &&
+          target::block_idx().x < static_cast<unsigned int>(arg.pack_blocks)) {
+        // first few blocks do packing kernel
+        typename Arg::template P<dslash.pc_type()> packer;
+        packer(arg, s, 1 - parity, dslash.twist_pack()); // flip parity since pack is on input
+
+        // we use that when running the exterior -- this is either
+        // * an explicit call to the exterior when not merged with the interior or
+        // * the interior with exterior_blocks > 0
 #ifdef NVSHMEM_COMMS
-    } else if (arg.shmem > 0
-               && ((kernel_type == EXTERIOR_KERNEL_ALL && arg.exterior_blocks == 0)
-                   || (kernel_type == UBER_KERNEL && arg.exterior_blocks > 0
-                       && blockIdx.x >= (gridDim.x - arg.exterior_blocks)))) {
-      shmem_exterior<kernel_type, D<nParity, dagger, xpay, kernel_type, Arg>, Arg, nParity>(dslash, arg, s);
+      } else if (arg.shmem > 0
+                 && ((kernel_type == EXTERIOR_KERNEL_ALL && arg.exterior_blocks == 0)
+                     || (kernel_type == UBER_KERNEL && arg.exterior_blocks > 0
+                         && target::block_idx.x >= (target::grid_dim().x - arg.exterior_blocks)))) {
```

Comment:
```suggestion
                         && target::block_idx().x >= (target::grid_dim().x - arg.exterior_blocks)))) {
```
This makes me wonder if GK has been tested with nvshemem turned on?

Reply:
I have tested it, but it's been a month or two.  I'll make sure to do that again this week.

Reply:
Compilation fixed now e4a1073a63a7063aed02f2d7e227b2b720fe729a